src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Map.cs

```diff
@@ -38,14 +41,199 @@ public void WriteEndMap()
+                CborConformanceLevel.Ctap2Canonical => true,
+                CborConformanceLevel.Strict => false,
+                CborConformanceLevel.NonStrict => false,
+                _ => false,
+            };
+        }
+
+        private SortedSet<(int offset, int keyLength, int keyValueLength)> GetKeyValueEncodingRanges()
+        {
+            // TODO consider pooling set allocations?
+
+            if (_keyValueEncodingRanges == null)
+            {
+                _comparer ??= new KeyEncodingComparer(this);
+                return _keyValueEncodingRanges = new SortedSet<(int offset, int keyLength, int keyValueLength)>(_comparer);
+            }
+
+            return _keyValueEncodingRanges;
+        }
+
+        private void HandleKeyWritten()
+        {
+            Debug.Assert(_currentKeyOffset != null && _currentValueOffset == null);
+
+            _currentValueOffset = _offset;
+
+            if (ConformanceRequiresUniqueKeys())
+            {
+                SortedSet<(int offset, int keyLength, int keyValueLength)> ranges = GetKeyValueEncodingRanges();
+
+                (int offset, int keyLength, int valueLength) currentKeyRange =
+                    (_currentKeyOffset.Value,
+                     _currentValueOffset.Value - _currentKeyOffset.Value,
+                     0);
+
+                if (ranges.Contains(currentKeyRange))
+                {
+                    // TODO: check if rollback is necessary here
+                    throw new InvalidOperationException("Duplicate key encoding in CBOR map.");
+                }
+            }
+        }
+
+        private void HandleValueWritten()
+        {
+            Debug.Assert(_currentKeyOffset != null && _currentValueOffset != null);
+
+            if (ConformanceRequiresUniqueKeys() || ConformanceRequiresSortedKeys())
+            {
+                SortedSet<(int offset, int keyLength, int keyValueLength)> ranges = GetKeyValueEncodingRanges();
+
+                (int offset, int keyLength, int keyValueLength) currentKeyRange =
+                    (_currentKeyOffset.Value,
+                     _currentValueOffset.Value - _currentKeyOffset.Value,
+                     _offset - _currentKeyOffset.Value);
+
+                ranges.Add(currentKeyRange);
+            }
+
+            // reset state
+            _currentKeyOffset = _offset;
+            _currentValueOffset = null;
+        }
+
+        private void SortKeyValuePairEncodings()
+        {
+            if (_keyValueEncodingRanges == null)
+            {
+                return;
+            }
+
+            int totalMapPayloadEncodingLength = _offset - _frameOffset;
+            byte[] tempBuffer = s_bufferPool.Rent(totalMapPayloadEncodingLength);
+            Span<byte> tmpSpan = tempBuffer.AsSpan(0, totalMapPayloadEncodingLength);
+
+            // copy sorted ranges to temporary buffer
+            Span<byte> s = tmpSpan;
+            foreach((int, int, int) range in _keyValueEncodingRanges)
```

Comment:
Please turn this tuple into a named structure.  It's appearing in too many places.

Reply:
I don't think it's as bad. Is it maybe because I omitted labels in this particular instance?

Reply:
* This foreach
* _keyValueEncodingRanges declaration
* _keyValueEncodingRanges initialization
* GetKeyValueEncodingRanges return value
* HandleKeyWritten
* HandleValueWritten
* GetKeyEncoding
* GetValueEncoding
* KeyEncodingComparer declaration
* KeyEncodingComparer.Compare (x2)

That's a lot of places that need to change if you want to carry one more piece of data later.

As a type the element names should be PascalCased; the camelCase form should only appear as a deconstruction-declaration syntax.

Reply:
> That's a lot of places that need to change if you want to carry one more piece of data later.

I suppose though that's the inherent tension when using tuples, no? Assuming the tuple accumulates > 3 elements then certainly converting to a named struct might be worth the effort, but there doesn't seem to be a need for that, at this point. There are other instances in the same implementation where tuples are being used to encode ranges as well.